File: javascript/examples/keys.js
import { Account } from "@near-js/accounts";
import { JsonRpcProvider } from "@near-js/providers";
import { KeyPairSigner } from "@near-js/signers";
import { KeyPair } from "@near-js/crypto";
import { NEAR } from "@near-js/tokens";

import dotenv from "dotenv";

dotenv.config({ path: "../.env" });

const accountId = process.env.ACCOUNT_ID;

// Create a connection to testnet RPC
const provider = new JsonRpcProvider({
  url: "https://test.rpc.fastnear.com",
});

// Query keys with the provider
const keysProvider = await provider.viewAccessKeyList(accountId);
console.log(keysProvider);

// Create an Account object without a signer
const account = new Account(accountId, provider);

// Query keys
const accessKeys = await account.getAccessKeyList();
console.log(accessKeys);


// Create a signer and add it to the Account
const privateKey = process.env.PRIVATE_KEY;
const signer = KeyPairSigner.fromSecretKey(privateKey); // ed25519:5Fg2...

account.setSigner(signer);

// New keys
const fullKeyPair = KeyPair.fromRandom("ed25519");
const fnKeyPair = KeyPair.fromRandom("ed25519");

// Add a Full Access Key
await account.addFullAccessKey(
  fullKeyPair.getPublicKey().toString() // ed25519:2ASWc...
);

// Add Function Call Access Key
await account.addFunctionCallAccessKey({
  publicKey: fnKeyPair.getPublicKey(), // The new public key ed25519:2ASWc...
  contractId: "example-contract.testnet", // Contract this key is allowed to call (optional)
  methodNames: ["example_method"], // Methods this key is allowed to call (optional)
  allowance: NEAR.toUnits("0.25") // Gas allowance key can use to call methods (optional)
  }
);

console.log(`Added FAK ${fullKeyPair.toString()}`);
console.log(`Added FCK ${fnKeyPair.toString()}`);

const accessKeysNew = await account.getAccessKeyList();
console.log(accessKeysNew);

// Use the new FullAccess Key to delete the Function Call Key
const newSigner = KeyPairSigner.fromSecretKey(fullKeyPair.toString());
account.setSigner(newSigner)
await account.deleteKey(fnKeyPair.getPublicKey().toString());
await account.deleteKey(fullKeyPair.getPublicKey().toString());

File: rust/examples/keys.rs
use dotenv::from_filename;
use near_api::prelude::{Account, AccountId, NetworkConfig, Signer};
use near_crypto::SecretKey;
use near_primitives::account::{AccessKeyPermission, FunctionCallPermission};
use std::str::FromStr;

#[tokio::main]
async fn main() {
    from_filename("../.env").unwrap();
    let private_key_string = std::env::var("PRIVATE_KEY").unwrap();
    let account_id_string = std::env::var("ACCOUNT_ID").unwrap();

    let account_id: AccountId = account_id_string.parse().unwrap();
    let account = Account(account_id.clone());

    let private_key = SecretKey::from_str(&private_key_string).unwrap();
    let signer = Signer::new(Signer::secret_key(private_key)).unwrap();

    let network = NetworkConfig::testnet();

    // Get all access keys for the account
    let keys = account.list_keys().fetch_from(&network).await.unwrap();
    println!("{:?}", keys);

    // Add full access key
    let (new_full_private_key, txn) = account
        .add_key(AccessKeyPermission::FullAccess)
        .new_keypair() // Generate a new keypair
        .generate_secret_key() // Get the private key from the keypair
        .unwrap();

    let new_full_public_key = new_full_private_key.public_key().to_string();
    println!("{:?}", new_full_public_key);

    let add_full_key_result = txn
        .with_signer(signer.clone())
        .send_to(&network)
        .await
        .unwrap();
    println!("{:?}", add_full_key_result);

    // Add function call access key
    let new_function_call_key = AccessKeyPermission::FunctionCall(FunctionCallPermission {
        allowance: Some(250_000_000_000_000_000_000_000), // Gas allowance key can use to call methods (optional)
        receiver_id: "example-contract.testnet".to_string(), // Contract this key is allowed to call
        method_names: vec!["example_method".to_string()], // Methods this key is allowed to call
    });

    let (new_function_private_key, txn) = account
        .add_key(new_function_call_key)
        .new_keypair()
        .generate_secret_key()
        .unwrap();

    let new_function_public_key = new_function_private_key.public_key().to_string();
    println!("{:?}", new_function_public_key);

    let add_function_key_result = txn
        .with_signer(signer.clone())
        .send_to(&network)
        .await
        .unwrap();
    println!("{:?}", add_function_key_result);

    // Delete full access key
    let public_key_to_delete = new_full_public_key;
    let delete_full_key_result = account
        .delete_key(public_key_to_delete.parse().unwrap())
        .with_signer(signer.clone())
        .send_to(&network)
        .await
        .unwrap();
    println!("{:?}", delete_full_key_result);
}
